arbnode/message_pruner.go

```diff
@@ -0,0 +1,109 @@
+	"time"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rpc"
+
+	"github.com/offchainlabs/nitro/staker"
+	"github.com/offchainlabs/nitro/util/stopwaiter"
+
+	flag "github.com/spf13/pflag"
+)
+
+type MessagePruner struct {
+	stopwaiter.StopWaiter
+	transactionStreamer *TransactionStreamer
+	inboxTracker        *InboxTracker
+	staker              *staker.Staker
+	config              MessagePrunerConfigFetcher
+}
+
+type MessagePrunerConfig struct {
+	Enable               bool          `koanf:"enable"`
+	MessagePruneInterval time.Duration `koanf:"prune-interval" reload:"hot"`
+}
+
+type MessagePrunerConfigFetcher func() *MessagePrunerConfig
+
+var DefaultMessagePrunerConfig = MessagePrunerConfig{
+	Enable:               false,
+	MessagePruneInterval: time.Minute,
+}
+
+func MessagePrunerConfigAddOptions(prefix string, f *flag.FlagSet) {
+	f.Bool(prefix+".enable", DefaultMessagePrunerConfig.Enable, "enable message pruning")
+	f.Duration(prefix+".prune-interval", DefaultMessagePrunerConfig.MessagePruneInterval, "interval for running message pruner")
+}
+
+func NewMessagePruner(transactionStreamer *TransactionStreamer, inboxTracker *InboxTracker, staker *staker.Staker, config MessagePrunerConfigFetcher) *MessagePruner {
+	return &MessagePruner{
+		transactionStreamer: transactionStreamer,
+		inboxTracker:        inboxTracker,
+		staker:              staker,
+		config:              config,
+	}
+}
+
+func (m *MessagePruner) Start(ctxIn context.Context) {
+	m.StopWaiter.Start(ctxIn, m)
+	m.CallIteratively(m.prune)
+}
+
+func (m *MessagePruner) prune(ctx context.Context) time.Duration {
+	latestConfirmedNode, err := m.staker.Rollup().LatestConfirmed(
+		&bind.CallOpts{
+			Context:     ctx,
+			BlockNumber: big.NewInt(int64(rpc.FinalizedBlockNumber)),
+		})
+	if err != nil {
+		log.Error("error getting latest confirmed node: %w", err)
+		return m.config().MessagePruneInterval
+	}
+	nodeInfo, err := m.staker.Rollup().LookupNode(ctx, latestConfirmedNode)
+	if err != nil {
+		log.Error("error getting latest confirmed node info: %w", err)
+		return m.config().MessagePruneInterval
+	}
+	endBatchCount := nodeInfo.Assertion.AfterState.GlobalState.Batch
+	if endBatchCount == 0 {
+		return m.config().MessagePruneInterval
+	}
+	endBatchMetadata, err := m.inboxTracker.GetBatchMetadata(endBatchCount - 1)
+	if err != nil {
+		log.Error("error getting last batch metadata: %w", err)
+		return m.config().MessagePruneInterval
+	}
+	deleteOldMessageFromDB(endBatchCount, endBatchMetadata, m.inboxTracker.db, m.transactionStreamer.db)
+	return m.config().MessagePruneInterval
+}
+
+func deleteOldMessageFromDB(endBatchCount uint64, endBatchMetadata BatchMetadata, inboxTrackerDb ethdb.Database, transactionStreamerDb ethdb.Database) {
+	if endBatchCount > 1 {
+		err := deleteFromRange(inboxTrackerDb, sequencerBatchMetaPrefix, 1, endBatchCount-1)
```

Comment:
This could traverse many millions of empty cells, each creating a database (HD) access.
I think we should store in endBatchCount and endBatchMetadata in MessagePruner.
Then after pruning 1..X, next prun only needs to check e.g. X-1..Y
Two options for first prun after node is woken up::
1) first prune is always 1...Y
2) Store X to database after prune is done, so can always prune X..Y (if w had an unexpected shutdown it's possible many of those values will be empty - which is not a problem)
I now lean towards option 2.

Reply:
done using option 2.